mosec/errors.py

```diff
@@ -24,19 +24,60 @@
 implemented by users), the `ValidationError` should be raised.
 """
 
+from .protocol import HTTPStautsCode
 
-class EncodingError(Exception):
+
+class MosecError(Exception):
+    """Mosec basic exception."""
+
+    code = HTTPStautsCode.INTERNAL_ERROR
+    msg = "mosec error"
+
+
+class ClientError(MosecError):
+    """Client side error.
+
+    This error indicates that the server cannot or will not process the request
+    due to something that is perceived to be a client error. It will return the
+    details to the client side with
+    [HTTP 400](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400).
+    """
+
+    code = HTTPStautsCode.BAD_REQUEST
+    msg = "bad request"
+
+
+class ServerError(MosecError):
+    """Server side error.
+
+    This error indicates that the server encountered an unexpected condition
+    that prevented it from fulfilling the request. It will return the details
+    to the client side with
+    [HTTP 500](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/500).
+
+    Attention: be careful about the returned message since it may contain some
+    sensitive information. If you don't want to return the details, just raise
+    an exception that is not inherited from `mosec.errors.MosecError`.
+    """
+
+    code = HTTPStautsCode.INTERNAL_ERROR
+    msg = "internal error"
+
+
+class EncodingError(ServerError):
     """Serialization error.
 
     The `EncodingError` should be raised in user-implemented codes when
     the serialization for the response bytes fails. This error will set
     to status code to
     [HTTP 500](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/500)
-    in the response.
+    and show the details in the response.
     """
 
+    msg = "encoding error"
+
 
-class DecodingError(Exception):
+class DecodingError(ClientError):
```

Comment:
Why DecodingError is ClientError but EncodingError is not?

Reply:
Encoding encodes the data generated by the server side. The users have no control over that. Thus I think it's usually related to some bug or corner cases in the code.

Reply:
How about user implements their own ser/de, and their Decoding is buggy?

Reply:
The server side here means the developers using mosec to deploy their models. The client side means the end users call the model services to get the results.

I guess mosec internal bugs will also be treated as server side errors.

Reply:
Got it, this two should be decided by the developers using mosec in their forward codes, thanks!